File: android/rxflux-core/src/main/java/com/klefmusic/rxfluxcore/middleware/Epic.kt
package com.klefmusic.rxfluxcore.middleware

import com.klefmusic.rxfluxcore.ReduxStore
import com.klefmusic.rxfluxcore.models.Action
import com.klefmusic.rxfluxcore.withMiddleware
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.kotlin.merge

typealias Epic<S> = (actions: Observable<Action>, state: () -> S) -> Observable<Action>

fun <S> epic(epic: Epic<S>): Epic<S> = epic
fun <S> emptyEpic(): Epic<S> = epic { _, _ -> Observable.empty() }

fun <S> combineEpics(vararg epic: Epic<S>): Epic<S> =
    epic { actions, state -> epic.map { it(actions, state) }.merge() }

fun <S> ReduxStore<S>.withEpics(vararg epics: Epic<S>): ReduxStore<S> =
    withEpics(combineEpics(*epics))

fun <S> ReduxStore<S>.withEpics(epics: Epic<S>): ReduxStore<S> =
    withMiddleware { store ->
        epics(store.actions, store::currentViewState).doOnNext { store.dispatch(it) }
            .ignoreElements()
    }

File: iOS/mobileflux/mobileflux/middleware/Epic.swift
//
//  Epic.swift
//  mobileflux
//
//  Created by RORY KELLY on 23/09/2020.
//

import Foundation
import RxSwift

typealias Epic<S> = (_ actions: Observable<Action>, _ state: () -> S) -> Observable<Action>

func epic<S>(epic: @escaping Epic<S>)-> Epic<S> {
    return epic
}
func emptyEpic<S> () -> Epic<S> {
    return epic { _,_ in Observable.empty() }
}

func combineEpics<S>(epics: [Epic<S>])-> Epic<S> {
    return epic { actions, state in
        Observable.from(epics.map { $0(actions, state) }).merge()
        
    }
}
func combineEpics<S>(epics: Epic<S>...)-> Epic<S> {
    return combineEpics(epics: epics)
}


func epicMiddleware<S>(epics: Epic<S>...) -> Middleware<S> {
    return { actions, dispatch, state in
        let allEpics = combineEpics(epics: epics)
        return allEpics(actions,  state )
            .do(onNext: { dispatch($0) })
            .ignoreElements()
            .asObservable()
            .map { $0 as! S }
    }
     
}

